onnx->torch bug when no input passed

When we convert from onnx to torch, the converter calculates the expected number of inputs wrong.
Ex:
```
convert(onnx_model, 'torch')
```

We need to force the user to pass some test input along with the conversion to prevent this issue.
Ex:
```
convert(onnx_model, 'torch', data)
```
should work.

For now we can add an `assert` with a helpful error message

I should clarify that this is a _predict-time error_, not runtime error:

```
from hummingbird.ml import convert
import pandas as pd
import onnx # for load from string
s='hospital_LogisticRegression_Binary_l1_0.1_batch.onnx'

with open(s, 'rb') as binary_file:
    mod = binary_file.read()

hbout = convert( onnx.load_from_string(mod),'torch')
# no error

data = pd.read_csv('data.csv')
hbout.predict(data)
```
gives:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/hummingbird/hummingbird/ml/containers/_sklearn_api_containers.py", line 119, in predict
    return self._run(self._predict, *inputs)
  File "/root/hummingbird/hummingbird/ml/containers/_sklearn_api_containers.py", line 67, in _run
    return function(*inputs)
  File "/root/hummingbird/hummingbird/ml/containers/sklearn/pytorch_containers.py", line 192, in _predict
    return self.model.forward(*inputs)[0].cpu().numpy().ravel()
  File "/root/hummingbird/hummingbird/ml/_executor.py", line 73, in forward
    self._input_names
AssertionError: number of inputs or number of columns in the dataframe do not match with the expected number of inputs ['hematocrit', 'neutrophils', 'sodium', 'glucose', 'bloodureanitro', 'creatinine', 'bmi', 'pulse', 'respiration', 'number_of_issues', 'asthma', 'depress', 'dialysisrenalendstage', 'fibrosisandother', 'gender', 'hemo', 'irondef', 'malnutrition', 'pneum', 'psychologicaldisordermajor', 'psychother', 'secondarydiagnosisnonicd9', 'substancedependence', 'rcount']
```

But this works:
```
hbout = convert(onnx.load_from_string(mod),'torch', data)   # pass in some test data

hbout.predict(data)
```

--- a/hummingbird/ml/convert.py
+++ b/hummingbird/ml/convert.py
@@ -80,9 +80,14 @@ def _supported_backend_check_config(model, backend, extra_config):
     if (
         (backend == torch.jit.__name__ and not _is_onnx_model(model))
         or backend == tvm_backend
-        or (backend == onnx.__name__ and not _is_onnx_model(model))
+        or (backend == onnx.__name__ and not _is_onnx_model(model) or _is_onnx_model(model) and backend == torch.__name__)
     ) and constants.TEST_INPUT not in extra_config:
-        raise RuntimeError("Backend {} requires test inputs. Please pass some test input to the convert.".format(backend))
+        raise RuntimeError(
+            "Backend {} requires test inputs. \
+            Please pass some test input to the convert function.".format(
+                backend
+            )
+        )
 
 
 def _convert_sklearn(model, backend, test_input, device, extra_config={}):

--- a/tests/test_onnxml_binarizer_converter.py
+++ b/tests/test_onnxml_binarizer_converter.py
@@ -89,6 +89,22 @@ class TestONNXBinarizer(unittest.TestCase):
 
         self.assertRaises(RuntimeError, convert, onnx_ml_model, "onnx", X)
 
+    @unittest.skipIf(
+        not (onnx_ml_tools_installed() and onnx_runtime_installed()), reason="ONNXML test requires ONNX, ORT and ONNXMLTOOLS"
+    )
+    # if the user doesn't add test input when converting to torch, we should get an error
+    def test_onnx_binarizer_converter_raises_rt_no_input(self):
+        warnings.filterwarnings("ignore")
+        X = np.array([[1, 2, 3], [4, 3, 0], [0, 1, 4], [0, 5, 6]], dtype=np.float32)
+        model = Binarizer(threshold=0)
+        model.fit(X)
+
+        # generate test input
+        onnx_ml_model = convert_sklearn(model, initial_types=[("float_input", FloatTensorType_onnx(X.shape))])
+        onnx_ml_model.graph.node[0].attribute[0].name = "".encode()
+
+        self.assertRaises(RuntimeError, convert, onnx_ml_model, "torch")
+
 
 if __name__ == "__main__":
     unittest.main()

